### PR TITLE
Capture-solver-log

### PR DESCRIPTION
**Pull request by Xemnas0**
Asyncronously capture output of the solver and save it as a string.
Allows for both "msg" and "logPath" to be enabled.
Change applied only to class COIN_CMD

### CODE REVIEW

**Comment by Xemnas0** [comment]
@pchtsp could you check this please?

**Comment by pchtsp** [comment]
can you update with master again? the tests were failing for another reason. thanks.

**Comment by Xemnas0** [comment]
Ah I see, python 2.7 does not support type annotations

File: pulp/apis/coin_api.py [hunk 1]
```diff
@@ -176,31 +178,70 @@ def solve_CBC(self, lp, use_mps=True):
...
+                return_code = self.loop.run_until_complete(self.run_sub(args))
+                self.full_output = "\n".join(self.output)
+                self.loop.close()
+                return return_code
+
+        sub_process_runner = SubProcessRunner(print_to_stdout=self.msg, logPath=logPath)
+        return_code = sub_process_runner.run(args)
+
+        if return_code != 0:
+            print(return_code)
```
Comment (jerr0328) [inline pulp/apis/coin_api.py:234]: Should remove this, otherwise this can break standardized logging

File: pulp/apis/coin_api.py [hunk 2]
```diff
@@ -176,31 +178,70 @@ def solve_CBC(self, lp, use_mps=True):
...
         log.debug(self.path + cmds)
         args = []
         args.append(self.path)
         args.extend(cmds[1:].split())
-        cbc = subprocess.Popen(args, stdout=pipe, stderr=pipe, stdin=devnull)
-        if cbc.wait() != 0:
-            if pipe:
-                pipe.close()
+
+        class SubProcessRunner:
```
Comment (jerr0328) [inline pulp/apis/coin_api.py:189]: I'm not the maintainer of the project so this is just an opinion, but to make testing easier, this class could be split out at a higher level and tested independently, rather than having the class declared inline.

File: pulp/apis/coin_api.py [hunk 3]
```diff
@@ -176,31 +178,70 @@ def solve_CBC(self, lp, use_mps=True):
...
+
+        sub_process_runner = SubProcessRunner(print_to_stdout=self.msg, logPath=logPath)
+        return_code = sub_process_runner.run(args)
+
+        if return_code != 0:
+            print(return_code)
             raise PulpSolverError(
                 "Pulp: Error while trying to execute, use msg=True for more details"
                 + self.path
+                + "return_code={}\noutput={}".format(
```
Comment (jerr0328) [inline pulp/apis/coin_api.py:238]: Exception would be hard to read with return code having no space before it, maybe newline here? Or remove this? I'd also suggest rewording the full exception, because the suggestion to `use msg=True for more details` would print to stdout but if you capture the output via other means, it will mislead developers thinking that `msg=True` will give more details than already provided.

**Review by jerr0328 — COMMENTED** [review]
Thanks, this indeed does seem to solve #502, as the solver output would be in the `solver_output` variable, though I haven't tried running with this branch. I've put some comments since I noticed some things that might cause problems for us if merged as-is.
